### DISCUSSION TITLE
Telescope keybindings

### DISCUSSION THREAD

**Discussion by rwijtvliet** [LazyVim/LazyVim#2940] [Q&A]
I'm remapping the navigational keys, from hjkl to htns. This mapping should be as universal as possible. 

I have already filled the `lua/config/keymaps.lua` file with the needed lines:
```
vim.keymap.set("", "h", "h", { expr = false, silent = true, desc = "move left" })
vim.keymap.set("", "n", "v:count == 0 ? 'gj' : 'j'", { expr = true, silent = true, desc = "move down" })
vim.keymap.set("", "t", "v:count == 0 ? 'gk' : 'k'", { expr = true, silent = true, desc = "move up" })
vim.keymap.set("", "s", "l", { expr = false, silent = true, desc = "move right" })
```

Now, I'm aware that plugins can define their own mappings. And if a plugin uses n, t, or s for something, the navigational movement is masked by the action the plugin links to n, t, or s; the navigation is not triggered until the n, t, or s key is manually mapped to a no-operation for the plugin.

What I was assuming here: if a plugin does NOT use the n, t, or s key for an action, surely the navigational action will work inside the plugin, as long as it uses navigation. 

However, this seems to not always be the case. In telescope, for example, this is not the case, and apparently it has defined its own navigation with hjkl, Looking at the mapped keys:
![latest-screenshot](https://github.com/LazyVim/LazyVim/assets/4106013/ee1f6951-a0c7-4867-9d07-02441ff4bde8)
 
My question is - is this expected? Is it expected that a plugin defines its own shortcuts for movement, that I'd need to disable the k, j and l keys in those plugins, and that I need to set the n, t, and s keys for movement? It just seems a bit cumbersome/elaborate. Of course, if there is no way around it, I'll do it.

**Comment by dpetka2001** [comment]
Those keys are mapped to Telescope actions that do the equivalent of moving up/down for Neovim, but throughout the results that the floating window is showing (it's basically operating on each entry shown in the Telescope window). So, it's natural that they would have to map them explicitly as far as I know.

**Comment by rwijtvliet** [comment]
I see, thanks for that. I've succeeded in making a new mapping for up and down: 

```lua
return {
  "nvim-telescope/telescope.nvim",
  opts = function(_, opts)
    local actions = require("telescope.actions")
    if not opts.defaults then
      opts.defaults = {}
    end
    if not opts.defaults.mappings then
      opts.defaults.mappings = {}
    end
    if not opts.defaults.mappings.n then
      opts.defaults.mappings.n = {}
    end
    opts.defaults.mappings.n["n"] = actions.move_selection_next
    opts.defaults.mappings.n["t"] = actions.move_selection_previous
  end,
}
```

Great!

Now, if I may ask a follow-up question that I've stumbled over a few times now.

I want/need to extend the `opts` table, as I do inside the function above. The issue is that I need to verify if a table exists, before I can add to it. Which causes the very verbose triple `if...end` blocks.

**Is there a better way to do this?**

When I was using astronvim, there was a `astronvim.utils` modlule with a `extend_tbl` function, made exactly for this type of deep merge. Lazynvim does not provide something similar, am I correct?

**Reply by dpetka2001** [reply]
Pretty sure that was just a wrapper around `vim.tbl_deep_extend`. You can use this to do what you want. You could do something like `vim.tbl_deep_extend("force", opts.defaults or {}, { settings_you_want_here })`.

**Comment by rwijtvliet** [comment]
That worked, many thanks for your help
